# 01. Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [2]:
#Define path
path = r'C:\Users\astat\Documents\Instacart Basket Analysis'

In [3]:
#Import data sets
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# 03. Consistency checks - Exercise

In [4]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [5]:
#Checking for missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [6]:
df_prods.shape

(49693, 5)

In [7]:
#Creating subset with missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [8]:
df_prods.shape

(49693, 5)

In [9]:
#Creating subset without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [10]:
df_prods_clean.shape

(49677, 5)

In [11]:
#Checking for duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [12]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [13]:
#Creating subset without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [14]:
df_prods_clean_no_dups.shape

(49672, 5)

In [15]:
#Checking for mixed types
for col in df_prods_clean_no_dups.columns.tolist():
  weird = (df_prods_clean_no_dups[[col]].applymap(type) != df_prods_clean_no_dups[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods_clean_no_dups[weird]) > 0:
    print (col)

# 04. Consistency Checks - Task

# 04.1 Consistency Checks - poducts

In [16]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

The maximum price with 99,999 seems too high for the type of products Instacart is selling.

In [17]:
#Running outlier checks on price
df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices'] > 100]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

In [18]:
#Excluding outliers by marking them NaN
df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices'] >100, 'prices'] = np.nan

In [19]:
#Checking max price value
df_prods_clean_no_dups['prices'].max()

25.0

In [20]:
df_prods_clean_no_dups['prices'].describe()

count    49670.000000
mean         7.680437
std          4.199381
min          1.000000
25%          4.100000
50%          7.100000
75%         11.100000
max         25.000000
Name: prices, dtype: float64

In [21]:
#Checking for missing values
df_prods_clean_no_dups.isnull().sum()

product_id       0
product_name     0
aisle_id         0
department_id    0
prices           2
dtype: int64

In [22]:
#Creating subset without missing values
df_prods_clean_no_dups2 = df_prods_clean_no_dups[df_prods_clean_no_dups['prices'].isnull() == False]

In [23]:
#Checking max price value
df_prods_clean_no_dups['prices'].max()

25.0

In [24]:
df_prods_clean_no_dups2['prices'].describe()

count    49670.000000
mean         7.680437
std          4.199381
min          1.000000
25%          4.100000
50%          7.100000
75%         11.100000
max         25.000000
Name: prices, dtype: float64

# 04.2 Consistency checks - orders

In [25]:
#Checking for mixed types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [26]:
#Checking for missing values
df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

The only variable with null values is 'days_since_last_order'. These can be explained by customers placing multiple orders on the same day, which would mean no day has passed between orders.

I would not make any changes to the null values, as they do not represent missing values and are important for the analysis.

In [27]:
#Checking for duplicates
df_dups = df_ords[df_ords.duplicated()]

In [28]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

There are no duplicate values in the orders df. 

In [29]:
#Exporting df
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv')) 